text/0000-is-sorted.md

```diff
@@ -0,0 +1,272 @@
+C++'s [`std::is_sorted`](http://en.cppreference.com/w/cpp/algorithm/is_sorted),
+Go's [`sort.IsSorted`](https://golang.org/pkg/sort/#IsSorted),
+D's [`std.algorithm.sorting.is_sorted`](https://dlang.org/library/std/algorithm/sorting/is_sorted.html)
+and others. (Curiously, many (mostly) more high-level programming language –
+like Ruby, Javascript, Java, Haskell and Python – seem to lack such a function.)
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Possible documentation of the three new methods of `Iterator`:
+
+> ```rust
+> fn is_sorted(self) -> bool
+> where
+>     Self::Item: Ord,
+> ```
+> Checks if the elements of this iterator are sorted.
+>
+> That is, for each element `a` and its following element `b`, `a <= b`
+> must hold. If the iterator yields exactly zero or one element, `true`
+> is returned.
+>
+> ## Example
+>
+> ```rust
+> assert!([1, 2, 2, 9].iter().is_sorted());
+> assert!(![1, 3, 2, 4).iter().is_sorted());
+> assert!([0].iter().is_sorted());
+> assert!(std::iter::empty::<i32>().is_sorted());
+> ```
+> ---
+>
+> ```rust
+> fn is_sorted_by<F>(self, compare: F) -> bool
+> where
+>     F: FnMut(&Self::Item, &Self::Item) -> Ordering,
+> ```
+> Checks if the elements of this iterator are sorted using the given
+> comparator function.
+>
+> Instead of using `Ord::cmp`, this function uses the given `compare`
+> function to determine the ordering of two elements. Apart from that,
+> it's equivalent to `is_sorted`; see its documentation for more
+> information.
+>
+> ---
+>
+> ```rust
+> fn is_sorted_by_key<F, K>(self, f: F) -> bool
+> where
+>     F: FnMut(&Self::Item) -> K,
+>     K: Ord,
+> ```
+> Checks if the elements of this iterator are sorted using the given
+> key extraction function.
+>
+> Instead of comparing the iterator's elements directly, this function
+> compares the keys of the elements, as determined by `f`. Apart from
+> that, it's equivalent to `is_sorted`; see its documentation for more
+> information.
+>
+> ## Example
+>
+> ```rust
+> assert!(["c", "bb", "aaa"].iter().is_sorted_by_key(|s| s.len()));
+> assert!(![-2i32, -1, 0, 3].iter().is_sorted_by_key(|n| n.abs()));
+> ```
+
+The methods for `[T]` will have analogous documentations.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This RFC proposes to add the following three methods to `[T]` (slices) and `Iterator`:
+
+```rust
+impl<T> [T] {
+    fn is_sorted(&self) -> bool
+    where
+        T: Ord,
+    { ... }
+
+    fn is_sorted_by<F>(&self, compare: F) -> bool
+    where
+        F: FnMut(&T, &T) -> Ordering,
+    { ... }
+
+    fn is_sorted_by_key<F, K>(&self, f: F) -> bool
+    where
+        F: FnMut(&T) -> K,
+        K: Ord,
+    { ... }
+}
+
+trait Iterator {
+    fn is_sorted(self) -> bool
+    where
+        Self::Item: Ord,
+    { ... }
+
+    fn is_sorted_by<F>(mut self, compare: F) -> bool
+    where
+        F: FnMut(&Self::Item, &Self::Item) -> Ordering,
```

Comment:
What is the use case for `FnMut` here as opposed to `Fn`?

Reply:
`FnMut` is *less* restrictive than `Fn` in where-clauses

Reply:
Yes, I am aware. But just because you *can* do something doesn't mean you *ought* to do something. So I'm wondering if there are any good reasons why you should be able to pass in a stateful closure.

Reply:
Ah, okay. But this would apply to almost all libstd functions that take closures, then. I don't think any method uses an `Fn` bound when `FnMut` suffices. This also applies to all sort methods we already have, so consistency is another argument for `FnMut`.

Reply:
Yep, that's exactly the reason I chose `FnMut`: it's everywhere already. In particular: [`[T]::sort_by_key`](https://doc.rust-lang.org/stable/std/primitive.slice.html#method.sort_by_key).

Do you think I should mention this in the RFC? Maybe that would be a good idea to prevent future readers from asking the same question about `FnMut`...

Reply:
So I knew about the consistency reason beforehand, but I think that we should decide on this with a rationale more than *"this is what we've done in other places"*, unless there's some other place where rationale has been given(?).

Reply:
@centril why should a function deliberately choose a more restrictive `Fn` than it needs, unless it wants to open itself to implementation changes that may require those restrictions?

The rules for picking Fn traits are simple:
* If you require reentrancy and concurrency, ask for `Fn`.

* If you only need to call it once, ask for `FnOnce`.
* If you call it multiple times but do not require reentrancy, ask for `FnMut`.
* If you need to call it concurrently, ask for `Fn`.

...and I rather doubt the stdlib needs to keep itself open to the possibility of changing this function to run in multiple threads.

By requiring `Fn` you make it difficult for the user to do things like caching and memoization, forcing people to use RefCell and thereby deferring compile-time borrow checks to runtime.  There is no upside.

Reply:
@ExpHP great arguments, so FnMut it is.